lib/http2/hpack.c

```diff
@@ -109,7 +116,11 @@ static int contains_invalid_field_name_char(const char *s, size_t len)
     for (; len != 0; ++s, --len) {
         unsigned char ch = (unsigned char)*s;
         if (!valid_h2_field_name_char[ch]) {
-            return 1;
+            if (ch - 'A' < 26U) {
+                *err_desc = err_found_upper_case_in_header_name;
+                return 1;
+            }
+            *err_desc = soft_err_found_invalid_char_in_header_name;
```

Comment:
Ah, you are right. Sorry for the fuss.

I think the confusion came from the fact that the name of the function (`contains_invalid_field_name_char`) pretty much asserts that it is a function that returns a boolean value (note that `contains_invaild_field_value_char` works that way). However, `contains_invalid_field_name_char` is actually returning if the validation should be stopped.

I would appreciate it if you could adjust the name (or possibly the signature of the function as well) to avoid the confusion.

Personally, I would think of changing the function to `const char *validate_header_name(const char *, size_t)` that returns `NULL` if validation succeeds, or an error string if it fails. After obtaining the error string, you could do a pointer comparison to see if it is a hard error or a soft error.